Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeAddress.php

```diff
@@ -0,0 +1,107 @@
+ *
+ * - {@see ContentRepositoryId}
+ * - {@see WorkspaceName}
+ * - {@see DimensionSpacePoint} (not to be confused with the {@see Node::$originDimensionSpacePoint})
+ * - {@see NodeAggregateId}
+ *
+ * In combination the parts can be used to distinctly identify a single node.
+ *
+ * By using the content graph for the content repository
+ * one can build a subgraph with the right perspective to find this node:
+ *
+ *      $subgraph = $contentRepository->getContentGraph($nodeAddress->workspaceName)->getSubgraph(
+ *          $nodeAddress->dimensionSpacePoint,
+ *          VisibilityConstraints::withoutRestrictions()
+ *      );
+ *      $node = $subgraph->findNodeById($nodeAddress->aggregateId);
+ *
+ * @api
+ */
+final readonly class NodeAddress implements \JsonSerializable
+{
+    private function __construct(
+        public ContentRepositoryId $contentRepositoryId,
+        public WorkspaceName $workspaceName,
+        public DimensionSpacePoint $dimensionSpacePoint,
+        public NodeAggregateId $aggregateId,
+    ) {
+    }
+
+    public static function create(
+        ContentRepositoryId $contentRepositoryId,
+        WorkspaceName $workspaceName,
+        DimensionSpacePoint $dimensionSpacePoint,
+        NodeAggregateId $aggregateId,
+    ): self {
+        return new self($contentRepositoryId, $workspaceName, $dimensionSpacePoint, $aggregateId);
+    }
+
+    public static function fromNode(Node $node): self
+    {
+        return new self(
+            $node->contentRepositoryId,
+            $node->workspaceName,
+            $node->dimensionSpacePoint,
+            $node->nodeAggregateId
+        );
+    }
+
+    /**
+     * @param array<string, mixed> $array
+     */
+    public static function fromArray(array $array): self
+    {
+        return new self(
+            ContentRepositoryId::fromString($array['contentRepositoryId']),
+            WorkspaceName::fromString($array['workspaceName']),
+            DimensionSpacePoint::fromArray($array['dimensionSpacePoint']),
+            NodeAggregateId::fromString($array['aggregateId'])
+        );
+    }
+
+    public static function fromJsonString(string $jsonString): self
+    {
+        return self::fromArray(\json_decode($jsonString, true, JSON_THROW_ON_ERROR));
+    }
+
+    public function withAggregateId(NodeAggregateId $aggregateId): self
+    {
+        return new self($this->contentRepositoryId, $this->workspaceName, $this->dimensionSpacePoint, $aggregateId);
+    }
+
+    public function equals(self $other): bool
+    {
+        return $this->contentRepositoryId->equals($other->contentRepositoryId)
+            && $this->workspaceName->equals($other->workspaceName)
+            && $this->dimensionSpacePoint->equals($other->dimensionSpacePoint)
+            && $this->aggregateId->equals($other->aggregateId);
+    }
+
+    public function toJson(): string
+    {
+        return json_encode($this, JSON_THROW_ON_ERROR);
```

Comment:
```suggestion
        try {
            return json_encode($this, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            throw new \RuntimeException(sprintf('Failed to JSON-encode NodeAddress: %s', $e->getMessage(), 1715608338, $e);
        }
```
Otherwise we have to add a `@throws` annotation and than catch that in the upper layers. This way we deal with it here and make the error a true runtime exception (that should never occur)

Reply:
hmm i agree that we need some kind of concept for exceptions and annotations that will also be linted (see https://phpstan.org/blog/bring-your-exceptions-under-control) but i rather not start this now in this case with the node address ;)

Reply:
No need to start anything – but this is not only about PHPStan. A JsonException is to be handled at the location the json de/encoding happens IMO.
If we don't do that in other places, we should change those at some point. But I don't see any reason why we should not make this safer in new code!?

Reply:
in the end we probably want to unify all code. There will be no error thrown in this case never. So im fine with `return json_encode($this)` as well but i like to avoid boiler plate code where im not so sure if i like this pattern to be at every place :O

Reply:
> There will be no error thrown in this case never

Famous last words :)

There will be errors if some of the underlying objects failed to encode or contain invalid characters.
My point is not so much about unification, but about improving the DX:
```php
json_encode("\xB1\x31", JSON_THROW_ON_ERROR);
```
Leads to an exception that no mere mortal can relate to the NodeAddress:
```
Uncaught JsonException: Malformed UTF-8 characters, possibly incorrectly encoded
```
vs
```
Failed to JSON-encode NodeAddress: Malformed UTF-8 characters, possibly incorrectly encoded
```

But I'll rest my case and will push a follow-up if you disagree :)

Reply:
sorry for being so stubborn :O - i dont want to insist on my side as well and you have done the changes would be a shame to lose the work ;)